src/v/raft/append_entries_buffer.cc

```diff
@@ -0,0 +1,121 @@
+#include "raft/append_entries_buffer.h"
+
+#include "raft/consensus.h"
+#include "raft/types.h"
+
+#include <seastar/core/coroutine.hh>
+#include <seastar/core/do_with.hh>
+#include <seastar/core/loop.hh>
+#include <seastar/util/later.hh>
+#include <seastar/util/variant_utils.hh>
+
+#include <exception>
+#include <variant>
+#include <vector>
+namespace raft {
+
+append_entries_buffer::append_entries_buffer(
+  consensus& c, size_t max_buffered_elements)
+  : _consensus(c)
+  , _max_buffered(max_buffered_elements) {}
+
+ss::future<append_entries_reply>
+append_entries_buffer::enqueue(append_entries_request&& r) {
+    return ss::with_gate(_gate, [this, r = std::move(r)]() mutable {
+        // we normally do not want to wait as it would cause requests
+        // reordering. Reordering may only happend if we would wait on condition
+        // variable.
+
+        return _flushed
+          .wait([this] { return _requests.size() < _max_buffered; })
+          .then([this, r = std::move(r)]() mutable {
+              ss::promise<append_entries_reply> p;
+              auto f = p.get_future();
+              _requests.push_back(std::move(r));
+              _responses.push_back(std::move(p));
+              _enqueued.signal();
+              return f;
+          });
+    });
+}
+
+ss::future<> append_entries_buffer::stop() {
+    _enqueued.broken();
+    _flushed.broken();
+    return _gate.close();
+}
+
+void append_entries_buffer::start() {
+    (void)ss::with_gate(_gate, [this]() mutable {
+        return ss::do_until(
+          [this] { return _gate.is_closed(); },
+          [this]() mutable {
+              return _enqueued.wait([this] { return !_requests.empty(); })
+                .then([this]() mutable { return flush(); })
+                .handle_exception_type(
+                  [](const ss::broken_condition_variable&) {
+                      // ignore exception, we are about to stop
+                  });
+          });
+    });
+}
+
+ss::future<> append_entries_buffer::flush() {
+    auto requests = std::exchange(_requests, {});
+    auto response_promises = std::exchange(_responses, {});
```

Comment:
Is it worth re-using the buffer for in-flight request to reduce allocs, at the cost of more average memory usage? I.e., swapping them here.